### PR TITLE
FlameGraph: Add docs for flame graph

### PR DESCRIPTION
**Pull request by joey-grafana**
**What is this feature?**

This adds documentation in the [visualizations section](https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/) of our docs.

### CODE REVIEW

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+## Flame graph
+
+The flame graph takes advantage of the hierarchical nature of profiling data. It aims to condense this data down into a format that allows you to easily see which code paths are consuming the most resources in your system.
+
+The way in which these resources are measured are through profiles which aggregate that information into a format which is then sent to the flame graph visualization. For example, allocated objects or space when measuring memory.
+
+{{< figure src="/static/img/docs/flame-graph-panel/flame-graph.png" class="docs-image--no-shadow" max-width= "900px" caption="Figure 1 - Flame graph" >}}
+
+Figure 1 shows and illustrates what this data looks like when rendered. The top block is the total and is the root or parent of all other blocks. It represents 100% of the profile(s) being measured. There may be 1 or more blocks for each subsequent row that is rendered below the total. Each of these blocks represents a child in the stack of function calls that originated from the original call to the root at the top of the flamegraph.
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:25]:
```suggestion
Figure 1 illustrates what this data looks like when rendered. The top block is the root or parent of the other blocks, and represents the total of all of the profiles being measured. There may be one or more blocks for each row that is rendered below the total. Each of these blocks represents a child in the stack of function calls that originated from the original call to the root at the top of the flame graph.
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+Each block can have 0 or more siblings but can only have one parent.
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:27]:
```suggestion
Each block can have zero or more siblings but can only have one parent.
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+In figure 1 we have many rows of colored blocks but also have grayed out sections which represent a set of blocks that represent a relatively short execution time and so were collapsed together into one section for performance reasons.
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:29]:
```suggestion
In figure 1 displays many rows of colored blocks as well as grayed-out sections which are a set of blocks that represent a relatively short execution time and were collapsed together into one section for performance reasons.
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+{{< figure src="/static/img/docs/flame-graph-panel/flame-graph-hovering.png" class="docs-image--no-shadow" max-width= "900px" caption="Figure 2 - Hovering over flame graph" >}}
+
+As there is more information to a block other than just its label, we decided to hide this information by default and show it on hover.
+
+Hovering over a specific block pops up a tooltip that shows you what the profile is measuring, in this case total time, along with other information such as this block's % of total time, the amount of time it took and the samples that were measured.
+
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:36]:
```suggestion
You can hover a specific block to view a tooltip that shows you what the profile is measuring, in this case total time, along with other information such as this block's % of total time, the amount of time it took, and the samples that were measured.

```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+{{< figure src="/static/img/docs/flame-graph-panel/flame-graph-clicking.png" class="docs-image--no-shadow" max-width= "900px" caption="Figure 3 - Clicking on a block" >}}
+
+As there may be multiple blocks per row, we may want to focus on a particular block in the row and its children only to further drill down the stack of function calls.
+
+Clicking on a particular block will do just that, hiding the block's siblings. The clicked block is set to 100% of the flame graphs width and all its children blocks are now shown with their widths updated relative to their new parents width.
+
+This process can be repeated to focus on finding the amount of resources that particular blocks and their children are consuming.
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:43]:
```suggestion
You can click on a block to drill down to the stack of function calls for that block and its child blocks. The clicked block is set to 100% of the flame graph's width and all its children blocks are shown with their widths updated relative to the width of the parent block.

This process can be repeated to focus on finding the amount of resources that particular blocks and their children are consuming.
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+{{< figure src="/static/img/docs/flame-graph-panel/flame-graph-search.png" class="docs-image--no-shadow" max-width= "900px" caption="Figure 4 - Search" >}}
+
+When you would like to quickly find the blocks with particular labels you can use the search field. Simply type in the label you want to search for and all of the blocks in the table with that text in their label will remain colored with all of the remaining blocks grayed out.
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:47]:
```suggestion
You can use the search field to find blocks with particular labels. When you search by a label name, all of the blocks in the table with labels that include the search text remain colored, while the rest of the blocks are grayed-out.
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+You can also click on any block or hover over it to narrow your search.
+
+## Top table
+
+The top table allows you to visualize which symbols take up the most resources in your profile.
+
+The table has three columns, symbols, self and total - representing the self time and total time of each symbol. The table is sorted by self time by default, but can be reordered by total time or symbol name upon clicking the column headers.
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:55]:
```suggestion
The table has three columns: symbols, self, and total. The table is sorted by self time by default, but can be reordered by total time or symbol name by clicking the column headers.
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+{{< figure src="/static/img/docs/flame-graph-panel/top-table.png" class="docs-image--no-shadow" max-width= "900px" caption="Figure 5 - Top table" >}}
+
+Clicking on a particular row of the top table will add that row's symbol to the search input which will search for that symbol in the flame graph.
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:59]:
```suggestion
Clicking on a row of the top table will add or remove that row's symbol to the search input for the flame graph.
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+Clicking on that top table row again will remove its symbol from the search input.
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:61]:
```suggestion

```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+{{< figure src="/static/img/docs/flame-graph-panel/top-table-clicking.png" class="docs-image--no-shadow" max-width= "900px" caption="Figure 6 - Clicking on a top table row" >}}
+
+Of course, sometimes you will not need to view both the top table and flame graph at the same time. This is especially true for large profiles.
+
+In this case and if you only want to see the top table, you can click on the top table button to show only the top table.
+
+3 options are available:
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:69]:
```suggestion
You can choose from three table view options:
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+- Top Table: Only show the top table
+- Flame Graph: Only show the flame graph
+- Both: Show both the top table and flame graph
+
+{{< figure src="/static/img/docs/flame-graph-panel/selected-view.png" class="docs-image--no-shadow" max-width= "900px" caption="Figure 7 - Show the top table / flame graph / both" >}}
+
+## Data API
+
+The flame graph expects a data frame data in a particular format to be able to render it. As the data is hierarchical while a data frame has a tabular structure, the way to encode it is to use a [nested set model](https://en.wikipedia.org/wiki/Nested_set_model).
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:79]:
```suggestion
In order to render the flame graph, you must format the data frame data using a [nested set model](https://en.wikipedia.org/wiki/Nested_set_model).
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+In a simplified form it means each item of the flame graph is encoded just by its nesting level as an integer value, its metadata and by its order in the data frame. This means that the order of items is significant and needs to be correct. The ordering is depth first traversal of the items in the flame graph. This allows us to recreate the graph without needing variable length values in the data frame like children's array.
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:81]:
```suggestion
A nested set model ensures each item of the flame graph is encoded just by its nesting level as an integer value, its metadata, and by its order in the data frame. This means that the order of items is significant and needs to be correct. The ordering is a depth-first traversal of the items in the flame graph which recreates the graph without needing variable-length values in the data frame like in a children's array.
```

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+Required fields:
```
Comment (Eve832) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:83]: Can we introduce this table?

**Review by Eve832 — APPROVED** [review]
Going to approve without mandating my suggested changes so I don't hold up the merge. Thank you!!

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+> **Note:** This panel is currently in beta & behind the `flameGraph` feature toggle.
+
+## Flame graph
+
+The flame graph takes advantage of the hierarchical nature of profiling data. It condenses data into a format that allows you to easily see which code paths are consuming the most system resources.
+
+These resources are measured through profiles which aggregate that information into a format which is then sent to the flame graph visualization. For example, allocated objects or space when measuring memory.
+
+{{< figure src="/static/img/docs/flame-graph-panel/flame-graph.png" class="docs-image--no-shadow" max-width= "900px" caption="Figure 1 - Flame graph" >}}
```
Comment (aocenas) [inline docs/sources/panels-visualizations/visualizations/flame-graph/index.md:23]: Shouldn't we use the markdown syntax for images? https://github.com/grafana/grafana/blob/main/contribute/documentation/documentation-markdown-guide.md#images

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+  - grafana
+  - dashboard
+  - documentation
+  - panels
+  - flame graph
+title: Flame graph
+weight: 850
+---
+
+# Flame graph panel
```
[the remark this answers is not in the document]
Comment (joey-grafana) [reply docs/sources/panels-visualizations/visualizations/flame-graph/index.md:15]: I removed the one below it :)

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md (hunk 13 above)
Comment (joey-grafana) [reply docs/sources/panels-visualizations/visualizations/flame-graph/index.md:79]: I'm not sure what you mean? There is already a table below this

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md (hunk 14 above)
Comment (joey-grafana) [reply docs/sources/panels-visualizations/visualizations/flame-graph/index.md]: Yeah! It looks like we have lots of other places to update this too :D

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md (hunk 9 above)
Comment (joey-grafana) [reply docs/sources/panels-visualizations/visualizations/flame-graph/index.md:57]: I updated the text as this is useful info for the user.

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md (hunk 14 above)
Comment (aocenas) [reply docs/sources/panels-visualizations/visualizations/flame-graph/index.md]: True I always see this elsewhere and going to copy it and then I remember that we should do it some other way.

File: docs/sources/panels-visualizations/visualizations/flame-graph/index.md (hunk 13 above)
Comment (aocenas) [reply docs/sources/panels-visualizations/visualizations/flame-graph/index.md:75]: I assume in the sense of giving it some description or why it's there? The whole section though is about what data shape does the flame graph require so I am not sure what would make sense here. Or do you mean just short `These are the required fields for flame graph:`?

**Pull request merged**